tensorboard/plugins/projector/vz_projector/data.ts

```diff
@@ -370,6 +366,92 @@ export class DataSet {
     });
   }
 
+  /** Runs UMAP on the data. */
+  async projectUmap(nComponents: number, nNeighbors: number, stepCallback: (iter: number) => void) {
+    this.hasUmapRun = true;
+    this.umap = new UMAP({nComponents, nNeighbors});
+
+    let nEpochs = 1;
+    let currentEpoch = 0;
+    const epochStepSize = 10;
+    const sampledIndices = this.shuffledDataIndices.slice(0, UMAP_SAMPLE_SIZE);
+  
+    const sampledData = sampledIndices.map(i => this.points[i]);
+    // TODO: Switch to a Float32-based UMAP internal
+    const X = sampledData.map(x => Array.from(x.vector));
+
+    this.nearest = await this.computeKnn(sampledData, nNeighbors);
+    
+    nEpochs = await util.runAsyncTask('Initializing UMAP...', () => {
+      const knnIndices = this.nearest.map(row => row.map(entry => entry.index));
+      const knnDistances = this.nearest.map(row => row.map(entry => entry.dist));
+
+      // Initialize UMAP and return the number of epochs.
+      return this.umap.initializeFit(X, knnIndices, knnDistances);
+    }, UMAP_MSG_ID);
+    
+    // Now, iterate through all epoch batches of the UMAP optimization, updating
+    // the modal window with the progress rather than animating each step since
+    // the UMAP animation is not nearly as informative as t-SNE.
+    return new Promise((resolve, reject) => {
+      const step = () => {
+        // Compute a batch of epochs since we don't want to update the UI
+        // on every epoch.
+        const epochsBatch = Math.min(epochStepSize, nEpochs - currentEpoch);
+        for (let i = 0; i < epochsBatch; i++) {
+          currentEpoch = this.umap.step();
+        }
+        const progressMsg = `Optimizing UMAP (epoch ${currentEpoch} of ${nEpochs})`;
+        
+        // Wrap the logic in a util.runAsyncTask in order to correctly update
+        // the modal with the progress of the optimization.
+        util.runAsyncTask(progressMsg, () => {
+        }, UMAP_MSG_ID, 0).then(() => {
```

Comment:
It's effectively the same thing (`runAsyncTask(noop).then(work, errorCB)` vs `runAsyncTask(work).then(noop, errorCB)`) and I had just copied over the success/error callback structure  from the t-SNE implementation. I've refactored to put the work in the `runAsyncTask` callback and use a `catch` call instead of the success/error pattern, which I think is now more clear.